### ISSUE TITLE
Faulty type variable unification in higher order function

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#5838]
With pyright 1.1.324 and the following code:

```py
from typing import Callable, TypeVar

A = TypeVar("A")
B = TypeVar("B")

def use_fn(fn: Callable[[A, B], A], b: B) -> A: ...
def fn(a: A, x: A) -> A: ...
def apply(a: A) -> A:
    return use_fn(fn, a)
```

I get the error:

```
error: Expression of type "A@apply | A@fn" cannot be assigned to return type "A@apply"
```
I would have expected this code to type check just fine as providing `fn` to `use_fn` should basically substitute the `B`s with more `A`s. Indeed:

```py
def use_fn_(fn: Callable[[A, A], A], b: A) -> A: ...
def apply_(a: A) -> A:
    return use_fn_(fn, a)
```
type checks.

Both versions work fine in mypy 1.5.1

**Comment by erictraut** [comment]
Thanks for the bug fix. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.325, which I just published. It will also be included in an upcoming pylance release.
